### PR TITLE
Add MetaTileEntity Solar Factory

### PR DESCRIPTION
**Pull request by purebluez**
image
idk how i should format this pr but it adds a multiblock called the Solar Factory.
solar factory can make solar cells, which can be made into solar panels in the same machine. 
these solar cells are cheaply upgradable (cheaper than dire autocrafting tables), but can also be made at twice the cost as a normal solar panel in exchange for skipping all the recursive crafting before it.
the solar factory can be used to navigate around the hard to use dire autocrafting tables, while making things cheaper ! 

image attached is a picture of the multiblock, its symmetrical.
lmk if you want more information !

### CODE REVIEW

**Comment by chochem** [comment]
> lmk if you want more information !

yes, you need expand on the PR description. not much there in terms of details.

**Comment by Dream-Master** [comment]
@purebluez yes show us a bit more

**Comment by purebluez** [comment]
> yes, you need expand on the PR description. not much there in terms of details.

alright!
- The multiblock:
    made from solid steel machine casings, niobium-titanium frame boxes, glass, and assembling line casings (the gearbox ones)
    can be rotated in all directions (idek if you can change this)
- The controller:
    made from ZPM parts (zpm hull, zpm motors, naquadria, indalloy solder) and various other materials
    made in a zpm assembly line, requiring a high pressure solar boiler as the research item 
- How exactly it affects balance:
    it makes tech solars (the solar panels that dont use any thaumcraft) quite a bit easier to craft, by providing resource discounts and not requiring the dire crafting tables.
    ive always felt that tech solars have the short end of the stick when compared to the magic solars, which find Much more use by the playerbase, because they are created from super cheap LV materials. this multi exists to offset the imbalance which exists between these two solar types.
    tech solars are additionally used in the eye of harmony casings, requiring the player to set up all of the dire autocrafting tables in order to autocraft EOH, ive heard from many endgame players on how annoying it is to set these dire autocrafting tables up, and the solar factory can craft the same solars while also being super easy to use !
- why its in zpm tier
    as it is, zpm is kind of content starved. the most important things in that tier are more fusion, naqline pt 2, that gt++ portal thing, and maybe some niche unlocks that the average player doesnt even see. 
    
i didnt know exactly what you wanted more of, so i gave basically everything

**Comment by boubou19** [comment]
Magic solar panels are a challenge of automation. If you want them in decent quantities, you need to get a good infra, as well as a getting the raw materials in quantities. Compared to tech solars, they are much more tedious to craft. Except Dyson, Solars have always been gated farther than any other alternative, to the point they aren't relevant. Single blocks with no maintenance costs being able to provide the same amount of energy as a machine that needs a whole line of process to be powered is an issue. 

If the dire crafting table is an issue for crafting, we can always make alternative recipes for when you actually need to spam GT solars in quantities, or we could use this multi to achieve that with no issue.

But really, @Dream-Master we must make sure that this doesn't make the solars top tier powergen again. Remember that you nerfed many times those solars to avoid exactly that (altho LuV+ should be reworked, they are too pricey for what they does, compared to the alternatives, but it's not the topic of this ticket). We have not planned to add this, so if we accept it as is we need someone to check if it doesn't break the powergen equilibrium. And that can take time.

Tldr: i'm against this change because it comes out of nowhere for a powergen that defeats the spirit of automation that is NH.

**Comment by Alastors** [comment]
Tech solars should get the short end of the stick compared to magic solars, like boubou said: they're significantly more interesting and challenging to craft than tech solars.

Also FFS mate, I thought my commit names were bad, you'd genuinely be better off leaving the names blank 🤦

File: src/main/java/gregtech/api/enums/GT_Values.java [hunk 1]
```diff
@@ -594,6 +594,17 @@ public static final class NBT {
         + EnumChatFormatting.BOLD
         + "0";
 
+    public static final String AuthorPureBluez = "Author: " + EnumChatFormatting.WHITE
```
Comment (Alastors) [inline src/main/java/gregtech/api/enums/GT_Values.java:597]: 💀

File: src/main/java/gregtech/api/enums/GT_Values.java [hunk 2]
```diff
@@ -594,6 +594,17 @@ public static final class NBT {
...
+        + "Pure"
+        + EnumChatFormatting.AQUA
```
Comment (Alastors) [inline src/main/java/gregtech/api/enums/GT_Values.java:599]: 💀

**Review by Alastors — CHANGES_REQUESTED** [review]

File: src/main/java/gregtech/api/enums/GT_Values.java [hunk 3]
```diff
@@ -594,6 +594,17 @@ public static final class NBT {
...
+        + "B"
+        + EnumChatFormatting.DARK_AQUA
+        + "l"
+        + EnumChatFormatting.BLUE
+        + "u"
+        + EnumChatFormatting.DARK_BLUE
+        + "ez";
```
Comment (Alastors) [inline src/main/java/gregtech/api/enums/GT_Values.java:606]: 💀

**Review by Alastors — DISMISSED** [review]

**Comment by DragonStorm25** [comment]
If this multi is really supposed to be a solar factory, it looks nothing like it. If anything, it looks like a lower tech version of the plasma mixer. I would suggest both a structure change and a block palette change. So many multis are made of the same gray blocks and glass, it would be refreshing to see new blocks. 

I have no comment on the balance as I have never used solars for power.

**Comment by EnderProyects** [comment]
If what you want is to address the issue of using the direct crafting table, it is tedious, you simply make the multiblock make the same recipe, but without giving more items than the original recipe, thus saving you progression issues...
Or since you mention this, why not an automatable version of the direcrafting table, which is capable of using different inputs for different recipes?

**Comment by Dream-Master** [comment]
> > yes, you need expand on the PR description. not much there in terms of details.
> 
> alright!
> 
> * The multiblock:
>   made from solid steel machine casings, niobium-titanium frame boxes, glass, and assembling line casings (the gearbox ones)
>   can be rotated in all directions (idek if you can change this)
> * The controller:
>   made from ZPM parts (zpm hull, zpm motors, naquadria, indalloy solder) and various other materials
>   made in a zpm assembly line, requiring a high pressure solar boiler as the research item
> * How exactly it affects balance:
>   it makes tech solars (the solar panels that dont use any thaumcraft) quite a bit easier to craft, by providing resource discounts and not requiring the dire crafting tables.
>   ive always felt that tech solars have the short end of the stick when compared to the magic solars, which find Much more use by the playerbase, because they are created from super cheap LV materials. this multi exists to offset the imbalance which exists between these two solar types.
>   tech solars are additionally used in the eye of harmony casings, requiring the player to set up all of the dire autocrafting tables in order to autocraft EOH, ive heard from many endgame players on how annoying it is to set these dire autocrafting tables up, and the solar factory can craft the same solars while also being super easy to use !
> * why its in zpm tier
>   as it is, zpm is kind of content starved. the most important things in that tier are more fusion, naqline pt 2, that gt++ portal thing, and maybe some niche unlocks that the average player doesnt even see.
> 
> i didnt know exactly what you wanted more of, so i gave basically everything

Steel machine casings does not fit in zpm tier . 

I will look into the machine ingame because I dindt see end reipe here shown .

File: src/main/java/gregtech/common/items/GT_MetaGenerated_Item_03.java [hunk 4]
```diff
@@ -916,6 +916,86 @@ public GT_MetaGenerated_Item_03() {
...
 
         ItemList.Timepiece.set(addItem(757, "Timepiece", "Beware of the kid with the hat", o));
+        ItemList.Solar_Cell.set(
+            addItem(
+                232,
+                "Solar Cell",
+                "Basic Solar Cell",
+                new TC_Aspects.TC_AspectStack(TC_Aspects.ELECTRUM, 64L),
+                new TC_Aspects.TC_AspectStack(TC_Aspects.POTENTIA, 64L),
+                new TC_Aspects.TC_AspectStack(TC_Aspects.TENEBRAE, 64L)));
```
Comment (BlueWeabo) [inline src/main/java/gregtech/common/items/GT_MetaGenerated_Item_03.java:926]: if you are going to put the same aspects into all solar cells, then do not even add aspects. If you want them at least make them different in some way. a basic solar cell should not give the same amount as a ultimate solar cell

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_SolarFactory.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    @Override
+    protected GT_Multiblock_Tooltip_Builder createTooltip() {
+        final GT_Multiblock_Tooltip_Builder tt = new GT_Multiblock_Tooltip_Builder();
+        tt.addMachineType("Solar Factory")
+            .addInfo("Cheaply & Quickly mass produces Solar Cells")
+            .addInfo("Has discounted recipes for Solar Panels")
+            .addInfo("Can upgrade Solar Cells to a higher tier with a")
+            .addInfo("cheap recipe.")
+            .addInfo("definitely not stolen from Tyrant")
```
Comment (BlueWeabo) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_SolarFactory.java:102]: This line of info is useless. It would be better suited for the multi instead of this

File: src/main/java/gregtech/api/enums/Textures.java [hunk 6]
```diff
@@ -1199,6 +1199,11 @@ public enum BlockIcons implements IIconContainer, Runnable {
         OVERLAY_FRONT_LARGE_CHEMICAL_REACTOR_ACTIVE,
         OVERLAY_FRONT_LARGE_CHEMICAL_REACTOR_ACTIVE_GLOW,
 
+        OVERLAY_FRONT_SOLAR_FACTORY_ACTIVE,
+        OVERLAY_FRONT_SOLAR_FACTORY_ACTIVE_GLOW,
+        OVERLAY_FRONT_SOLAR_FACTORY_GLOW,
+        OVERLAY_FRONT_SOLAR_FACTORY,
```
Comment (BlueWeabo) [inline src/main/java/gregtech/api/enums/Textures.java:1205]: if this is inactive, please say that. Sometimes its better to start being better than the old code

File: src/main/java/gregtech/loaders/postload/recipes/SolarFactoryRecipes.java [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+
+public class SolarFactoryRecipes implements Runnable {
+
+    @Override
+    public void run() {
+        // All the solar cells (1V - UV)
+        GT_Values.RA.stdBuilder()
+            .itemInputs(
+                getModItem(NewHorizonsCoreMod.ID, "item.ReinforcedGlassPlate", 2),
+                getModItem(NewHorizonsCoreMod.ID, "item.AluminiumIronPlate", 1),
```
Comment (BlueWeabo) [inline src/main/java/gregtech/loaders/postload/recipes/SolarFactoryRecipes.java:27]: Can you put all getModItem calls to have their own Itemstack? As most can be reused in other recipes thus caching the stack. Only thing you would need to do is maybe call a method from itemStack like ItemStack#splitItemStack iirc

File: src/main/java/gregtech/api/enums/ItemList.java [hunk 8]
```diff
@@ -814,6 +815,17 @@ public enum ItemList implements IItemContainer {
     Machine_HP_Compressor,
     Machine_HP_AlloySmelter,
 
+    Solar_Cell,
```
Comment (BlueWeabo) [inline src/main/java/gregtech/api/enums/ItemList.java:818]: if this is a blank solar cell say that or specify it.

**Review by BlueWeabo — COMMENTED** [review]
Overall did good for what you wanted, but I would've appreciated if there was something more to the Solar Factory other than a shape and a basic recipe map. Nothing much interesting happening with the multiblock, only it having Perfect Overclock, which I can't see it being needed currently as it doesn't cost anything.

**Comment by purebluez** [comment]
> Tech solars should get the short end of the stick compared to magic solars, like boubou said: they're significantly more interesting and challenging to craft than tech solars.
The best magic solar creates 600m eu/t from a thunderstorm, while the best tech solar creates only 524288 eu/t. that is 1144x more powergen because magic solars require thaumcraft. one of the many problems that tech solars have is that they dont scale past a uv amp, while magic solars keep going long after. There are currently solar panels that go up to UIV, but they dont have a recipe. This pr doesnt have a recipe for them either, but the solar factory would be a perfect way to create a recipe for them. 
Plus, I dont see the challenge from putting World Accelerators on an infusion altar and integrating the altar into ae2.

**Comment by boubou19** [comment]
Then i invite you to attempt making an octuple solar panel. Summing up magic solar panels simply to an accelerated and automated infusion altar is a great mistake. You clearly underestimate the infrastructure, power, materials and essentia required to make interesting compressed solar panels in a decent amount of time.

**Comment by chochem** [comment]
> while the best tech solar creates only 524288 eu/t.
no that is not true. There is the dyson swarm/sphere after that. Well currently there is a gap, but there are already detailed plans to rework it.

**Comment by Dream-Master** [comment]
you need to resolve the conflict 
https://github.com/purebluez/GT5u-SolarFactory/pull/1

**Comment by Dream-Master** [comment]
I have an idea. Make a multi for Solar is a nice idea but instead of the recipes you add why not using the original components with a discount and make the multi available in LV age. It use steel as casing so it can be produced in lv/mv or hv. Maybe make it like the pcb factory add some multi structure to make solar pannels cheaper or make it more efficient. Now the recipes looks very boring.

**Comment by Dream-Master** [comment]
image
this one need a icon and need to be added to nei

**Comment by Dream-Master** [comment]
image
backround for the crafting fiels looks a bit wired.

**Comment by Dream-Master** [comment]
The new icons are not necessary i would say and they looks more like some bords not like solar pannels

**Comment by Dream-Master** [comment]
maybe you can add more tiers of casings and frame boxes to make discounts if you dont want to add pcb style stuff. Or you add it like the ebf need different coils per tier. So steel casings and frame box can only make lv-hv pannels  titanium ev and so on

File: src/main/java/gregtech/common/items/GT_MetaGenerated_Item_03.java [hunk 9]
```diff
@@ -916,6 +916,86 @@ public GT_MetaGenerated_Item_03() {
         ItemList.Optically_Compatible_Memory.set(addItem(725, "Optically Compatible Memory", "Its in the name!", o));
 
         ItemList.Timepiece.set(addItem(757, "Timepiece", "Beware of the kid with the hat", o));
+        ItemList.Solar_Cell.set(
+            addItem(
+                232,
```
Comment (querns) [inline src/main/java/gregtech/common/items/GT_MetaGenerated_Item_03.java:921]: I put the Metrics Cover at ID 232 in MetaGenerated_Item_03. You'll probably have to adjust this.

Comment (Dream-Master) [reply src/main/java/gregtech/common/items/GT_MetaGenerated_Item_03.java:921]: @querns thats why i make the pr i mentioned earlier. Better use 240-249

**Comment by Dream-Master** [comment]
I have an additional idea. We could add the pannels to that multi to make recipes not recursive. but sure they need to be rebalanced.

**Comment by purebluez** [comment]
> I have an idea. Make a multi for Solar is a nice idea but instead of the recipes you add why not using the original components with a discount and make the multi available in LV age. It use steel as casing so it can be produced in lv/mv or hv. Maybe make it like the pcb factory add some multi structure to make solar pannels cheaper or make it more efficient. Now the recipes looks very boring.

love these ideas, currently working on 'porting' the solar factory MetaTileEntity to MuTEs to make this easier to implement. should i open a new PR on a branch like MuTEMaster once im done (aka closing this, and then opening one), or just make a fresh commit in this fork?

**Comment by Dream-Master** [comment]
@purebluez if it becomes a MuTE sure make a new pr to MuTEMaster.

**Comment by chochem** [comment]
I have to say ZPM makes a lot more sense than LV, you should stick with that.
Not only do solars start in late HV, but more importantly, discounted massproduction of solars would be incredibly unbalanced in early tiers and only makes sense in late tiers when the competition is powerful stuff like fusion/nq and maybe one needs many of them for dyson or eoh.

**Comment by Dream-Master** [comment]
@purebluez any news here ?

**Comment by purebluez** [comment]
> @purebluez any news here ?

I'm currently waiting on a MuTEMaster commit where the recipes work but I've fully made the MuTE structure and it works without fuss. even have a texture for one of the casings

**Comment by purebluez** [comment]
Gonna close this PR, due to a very long period of inactivity and that I wont be making any more updates to the Metatile version of Solar Factory. Im not really motivated to do more work on the MuTE version atm either, so I will make a PR in feature/MuTEMaster whenever I finish it. (no promises though)
